sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -63,10 +63,33 @@ template <typename Compare = std::less<>> class default_sorter {
   void operator()([[maybe_unused]] Group g, [[maybe_unused]] Ptr first,
                   [[maybe_unused]] Ptr last) {
 #ifdef __SYCL_DEVICE_ONLY__
+    // Adjust the scratch pointer based on alignment of the type T.
     // Per extension specification if scratch size is less than the value
     // returned by memory_required then behavior is undefined, so we don't check
     // that the scratch size statisfies the requirement.
-    sycl::detail::merge_sort(g, first, last - first, comp, scratch.data());
+    using T = typename sycl::detail::GetValueType<Ptr>::type;
+    size_t n = last - first;
+    void *scratch_ptr = scratch.data();
+    size_t space = scratch.size();
+    scratch_ptr = std::align(alignof(T), n * sizeof(T), scratch_ptr, space);
+    std::byte *aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+    const size_t idx = g.get_local_linear_id();
+    const size_t local = g.get_local_range().size();
+    T *local_copy = nullptr;
+    // Each work item constructs several elements based on the input size.
+    // Remember pointer to the beginning of the chunk in local_copy.
+    const size_t chunk = (n - 1) / local + 1;
+    for (int i = chunk; i >= 0; --i) {
+      auto index = idx * chunk + i;
```

Comment:
I think something like this would be clearer:

```
// Initialize in chunks: 
// WI  |   0              |  1                          | ...
// Idxs| 0..chunk_size -1 | chunk_size..2*chunk_size -1 | ...
for (int i = idx*chunk_size; i < (idx+1)*chunk_size -1; ++i)
```

or, if it would be more beneficial to the performance make it interleaved:
```
for (int i = idx; i < n; i+=local_size)
```

Or are you doing something completely different here?

Reply:
I am iterating backwards so that I save pointer to the first element in the chunk because I need to broadcast pointer to the beginning of the scratch later.